delocate/tools.py

```diff
@@ -812,22 +813,42 @@ def add_rpath(filename: str, newpath: str, ad_hoc_sign: bool = True) -> None:
         replace_signature(filename, "-")
 
 
-def zip2dir(zip_fname: str, out_dir: str) -> None:
+def zip2dir(
+    zip_fname: str | PathLike[str], out_dir: str | PathLike[str]
+) -> None:
     """Extract `zip_fname` into output directory `out_dir`
 
     Parameters
     ----------
-    zip_fname : str
+    zip_fname : str or Path-like
         Filename of zip archive to write
-    out_dir : str
+    out_dir : str or Path-like
         Directory path containing files to go in the zip archive
     """
-    # Use unzip command rather than zipfile module to preserve permissions
+    # The zipfile module does not preserve permissions correctly
     # http://bugs.python.org/issue15795
-    _run(["unzip", "-q", "-o", "-d", out_dir, zip_fname], check=True)
-
-
-def dir2zip(in_dir, zip_fname):
+    with zipfile.ZipFile(zip_fname, "r") as zip:
+        for name in zip.namelist():
+            member = zip.getinfo(name)
+            extracted_path = zip.extract(member, out_dir)
+            attr = member.external_attr >> 16
```

Comment:
How does this logic relate to the results of `unzip`?

Reply:
`zip.extract` does a pretty good at preserving file attributes, [but it doesn't do a perfect job due to its known issues with file permissions](http://bugs.python.org/issue15795).

After the file is extracted the rest of the code is to fix the permissions and preserve the timestamp.

Reply:
Are the relationships between the `external_attr` and desired output permissions written up somewhere that we could cite?   It would be good to read the code and see why permissions are in the most significant 2 bytes of the int, why directories are always 755, why 0 means leave permissions as they are...

Reply:
> Are the relationships between the external_attr and desired output permissions written up somewhere that we could cite?

Probably not, since when I try to look it up I find that `external_attr`'s poor documentation is a known issue:
https://github.com/python/cpython/issues/62462

I've been assuming theses are internals to the zip format itself.

Also not sure why directories are always 755, going back to auditwheel and following the blame logs for that code doesn't give an explanation.  So I just take it at face value as making sure the directory has enough permissions to continue extracting files and maybe running them.

Reply:
Maybe add comment and that link to let people know that this is an inherited heuristic, and in the hope that better documentation surfaces at some point.

Reply:
You mean `external_attr`?  It seems it's platform dependent too, with different platforms setting different bits, which explains why the high bits were compared to zero since a non-Unix platform probably won't even set them.

The most information I found on it was here: https://unix.stackexchange.com/questions/14705/the-zip-formats-external-file-attribute

Reply:
I've changed to better variable names and added a link to the most useful resource I could find on external_attr.